Mathlib/Analysis/Analytic/Prod.lean

```diff
@@ -0,0 +1,95 @@
+Authors: Geoffrey Irving
+-/
+import Mathlib.Analysis.Analytic.Composition
+import Mathlib.Analysis.Analytic.Linear
+
+/-!
+# Analytic functions on and to product spaces
+-/
+
+noncomputable section
+
+open Topology Classical BigOperators NNReal Filter ENNReal
+
+open Function (uncurry)
+
+open Set Filter Asymptotics
+
+variable {𝕜 E F G H : Type*}
+
+variable [NontriviallyNormedField 𝕜] [NormedAddCommGroup E] [NormedSpace 𝕜 E] [NormedAddCommGroup F]
+  [NormedSpace 𝕜 F] [NormedAddCommGroup G] [NormedSpace 𝕜 G] [NormedAddCommGroup H]
+  [NormedSpace 𝕜 H]
+
+/-- Products of analytic functions are analytic -/
+theorem AnalyticAt.prod {f : E → F} {g : E → G} {x : E} (fa : AnalyticAt 𝕜 f x)
+    (ga : AnalyticAt 𝕜 g x) : AnalyticAt 𝕜 (fun x ↦ (f x, g x)) x := by
+  rcases fa with ⟨p, pr, fp⟩
+  rcases ga with ⟨q, qr, gq⟩
+  set pq : FormalMultilinearSeries 𝕜 E (F × G) := fun n ↦ (p n).prod (q n)
+  have pqr : min pr qr ≤ pq.radius := by
+    apply ENNReal.le_of_forall_nnreal_lt; intro r rr
+    rcases p.norm_mul_pow_le_of_lt_radius
+        (lt_of_lt_of_le rr (_root_.trans (min_le_left pr qr) fp.r_le)) with
+      ⟨pc, _, ph⟩
+    rcases q.norm_mul_pow_le_of_lt_radius
+        (lt_of_lt_of_le rr (_root_.trans (min_le_right pr qr) gq.r_le)) with
+      ⟨qc, _, qh⟩
+    apply pq.le_radius_of_bound (max pc qc); intro n
+    calc ‖pq n‖ * ↑r^n
+      _ = max ‖p n‖ ‖q n‖ * ↑r^n := by simp only [ContinuousMultilinearMap.op_norm_prod]
+      _ = max (‖p n‖ * ↑r^n) (‖q n‖ * ↑r^n) := max_mul_of_nonneg _ _ (pow_nonneg (coe_nonneg r) n)
+      _ ≤ max pc qc := max_le_max (ph n) (qh n)
+  use pq, min pr qr
+  exact
+    { r_le := pqr
+      r_pos := lt_min fp.r_pos gq.r_pos
+      hasSum := by
+        intro y yr; apply HasSum.prod_mk
+        exact fp.hasSum (EMetric.ball_subset_ball (min_le_left _ _) yr)
+        exact gq.hasSum (EMetric.ball_subset_ball (min_le_right _ _) yr) }
+
+-- Products of analytic functions are analytic -/
+theorem AnalyticOn.prod {f : E → F} {g : E → G} {s : Set E} (fa : AnalyticOn 𝕜 f s)
+    (ga : AnalyticOn 𝕜 g s) : AnalyticOn 𝕜 (fun z ↦ (f z, g z)) s :=
+  fun _ m ↦ (fa _ m).prod (ga _ m)
+
+/-- `AnalyticAt.comp` for a curried function -/
+theorem AnalyticAt.curry_comp {h : F → G → H} {f : E → F} {g : E → G} {x : E}
+    (ha : AnalyticAt 𝕜 (uncurry h) (f x, g x)) (fa : AnalyticAt 𝕜 f x)
+    (ga : AnalyticAt 𝕜 g x) :
+    AnalyticAt 𝕜 (fun x ↦ h (f x) (g x)) x := by
+  have e : (fun x ↦ h (f x) (g x)) = uncurry h ∘ fun x ↦ (f x, g x) := rfl
+  rw [e];
+  exact AnalyticAt.comp ha (fa.prod ga)
+
+/-- `AnalyticOn.comp` for a curried function -/
+theorem AnalyticOn.curry_comp {h : F → G → H} {f : E → F} {g : E → G} {s : Set (F × G)} {t : Set E}
+    (ha : AnalyticOn 𝕜 (uncurry h) s) (fa : AnalyticOn 𝕜 f t) (ga : AnalyticOn 𝕜 g t)
+    (m : ∀ x, x ∈ t → (f x, g x) ∈ s) : AnalyticOn 𝕜 (fun x ↦ h (f x) (g x)) t := fun _ xt ↦
+  (ha _ (m _ xt)).curry_comp (fa _ xt) (ga _ xt)
```

Comment:
```suggestion
    (m : ∀ x, x ∈ t → (f x, g x) ∈ s) : AnalyticOn 𝕜 (fun x ↦ h (f x) (g x)) t := 
  fun _ xt ↦ (ha _ (m _ xt)).curry_comp (fa _ xt) (ga _ xt)
```
I think we tend not to keep the `fun` on the `:=` line, although probably Mathlib is inconsistent about this.

Reply:
Is there a way to lint for this?